x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextField.java

```diff
@@ -425,15 +425,16 @@ public static Chunk toSemanticTextFieldChunk(
         ChunkedInference.Chunk chunk,
         boolean useInferenceMetadataFieldsFormat
     ) {
-        // TODO: Use offsets from ChunkedInferenceServiceResults
-        // TODO: When using legacy semantic text format, build chunk text from offsets
-        assert chunk.matchedText() != null; // TODO: Remove once offsets are available from chunk
-        int startOffset = useInferenceMetadataFieldsFormat ? input.indexOf(chunk.matchedText()) + offsetAdjustment : -1;
-        return new Chunk(
-            useInferenceMetadataFieldsFormat ? null : chunk.matchedText(),
-            useInferenceMetadataFieldsFormat ? startOffset : -1,
-            useInferenceMetadataFieldsFormat ? startOffset + chunk.matchedText().length() : -1,
-            chunk.bytesReference()
-        );
+        String text = null;
+        int startOffset = -1;
+        int endOffset = -1;
+        if (useInferenceMetadataFieldsFormat) {
+            startOffset = chunk.textOffset().start() + offsetAdjustment;
+            endOffset = chunk.textOffset().end() + offsetAdjustment;
+        } else {
+            text = input.substring(chunk.textOffset().start(), chunk.textOffset().end());
```

Comment:
Can we also don't keep the `input` around if `useInferenceMetadataFieldsFormat` is true? I mean we can save some memory in `ShardBulkInferenceAction` by not keeping the input in such case.

Reply:
Sounds like a good optimization for a follow-up IMO. We need to keep the input in at least `FieldInferenceRequest` for actually performing inference, but we could conditionally set it in `FieldInferenceResponse`.

Reply:
sure